File: Striver/Day 25 : DP-1/MCM.cpp
//Question: https://practice.geeksforgeeks.org/problems/matrix-chain-multiplication0303/1

//Solution: class Solution{
public:
    int t[101][101];
    int solve (int arr[], int i, int j){
        if (i >= j)
            return 0;
        if(t[i][j] != -1)
            return t[i][j];
        int ans=INT_MAX;
        for (int k=i; k<j; k++){
            int temp = solve(arr,i,k) + solve(arr,k+1,j) + (arr[i-1]*arr[k]*arr[j]);
           // cout<<temp<<" "<<ans<<endl;
            ans = min(ans,temp);
        }
        return t[i][j] = ans;
    }
    
    int matrixMultiplication(int N, int arr[])
    {
        
        for (int i=0; i<101; i++){
            for (int j=0; j<101; j++){
                t[i][j] = -1;
            }
        }
        return solve(arr,1,N-1);
    }

File: Striver/Day 25 : DP-1/MCM.java
class Solution{
    static int matrixMultiplication(int N, int arr[])
    {
        // code here
        
        int[][] dp = new int[arr.length][arr.length];
        
        for(int[] arrr : dp)
          Arrays.fill(arrr , -1);
        
        return mcm(arr , 1 , arr.length-1 , dp);
        
    }
    
    public static int mcm(int[] arr , int i , int j , int[][] dp)
    {
        
        if(i>=j)
        return dp[i][j] = 0;
        
        if(dp[i][j] != -1)
        return dp[i][j];
        
        int ans = Integer.MAX_VALUE;
        
        for(int k = i ; k<=j-1 ; k++)
        {
            
            int left = mcm(arr , i , k ,dp);
            int right = mcm(arr , k+1 , j , dp);
            
            ans = Math.min(ans , left+right+arr[i-1]*arr[k]*arr[j]);
            
            
        }
    
        return dp[i][j] = ans;
    }
    
}

